### DISCUSSION TITLE
Is there a way to handle SIGTERM event in TSED lifecycle hook (Graceful Shutdown event)

### DISCUSSION THREAD

**Discussion by ashprojects** [TypedProject/ts-express-decorators#2586] [General]
We deploy our code to ECS containers. ECS would give us SIGTERM to have a graceful shutdown where we do following things:
1. Shutdown SQS Pollers
2. Shutdown External Polling (involves thirdparty)

In node we can easily catch SIGTERM with this:
```
const signalsToConsider: string[] = ["SIGTERM", "SIGINT"];
    for (const signal of signalsToConsider) {
      process.on(signal, () => {
        $log.info({event: "Signal received", signal, msg: "Calling hooks"});
        this._callHooks(); // can call hooks
      });
```
I read about Custom Event in Injector Service here: https://tsed.io/docs/hooks.html#subscribe-to-a-hook
But can't find module-emitter package anywhere. Currently on v6.101

It would be easy if I could raise an event (ex $onGracefulShutdown) and each corresponding service would implement and take action accordingly.

    }

**Comment by Romakita** [comment]
Hello @ashprojects 

If you generate a project using CLI, the index.ts will contains all of your needs.
Then, inside an injectable service, you just need to add `$onDestroy` method in your class and Ts.ED will call you
when the server instance is destroyed.

```ts
async function bootstrap() {
  try {
    const platform = await PlatformExpress.bootstrap(Server);
    await platform.listen();

    process.on("SIGINT", () => {
      platform.stop();
    });
  } catch (error) {
    $log.error({event: "SERVER_BOOTSTRAP_ERROR", message: error.message, stack: error.stack});
  }
}

bootstrap();

```

Also, I just that all of explanation are is the hook documentation. Maybe something is not clear, or you have missed the first example in the doc.

See you

**Reply by Romakita** [reply]
```ts
@Injectable()
class MyInjectableService {
  $onDestroy() {
    // do something
  }
}

**Reply by ashprojects** [reply]
Got ya.
But:

SIGTERM is a graceful shutdown signal. I don't think our services can be destroyed as it would kill partially completed async tasks. Generally there is a delay from SIGTERM to SIGKILL (which means kill the process). In that duration server can gracefully shutdown picking new things from Queues / etc and finish the current work

**Reply by ashprojects** [reply]
I think my use case is more specific, raise a custom event on SIGTERM and it becomes available in Service as $onGracefulShutdown so that they can run the logic. Probably

**Reply by Romakita** [reply]
Why? If sigterm is emitted, calling platform.stop() will emit $onDestroy event. It’s exactly what you need.

**Reply by Romakita** [reply]
So instead binding stop to SIGINT bind the method to SIGTERM. The idea of $onDestroy hook is to close gracefully all db connections and other stuff.

see you
